Improve R Tracking API

kevinykuo opened the pull request:
This PR exposes the previously unexported tracking client API and attains parity with the REST and Python interfaces.

The "fluent" and "client" features are unified into one set of functions. When provided, the client object is be respected (similar to Python's `mlflow.tracking`). For functions concerned with runs, an active run is started and registered if `run_id` is not specified (similar to the fluent API).

Data structures returned by tracking functions are standardized to be tidy tibbles. Each experiment is a row, each run is a row, and each data element record (metric/param/tag) is a row. This enables easier inspection and further manipulation compared to nested lists.

## Examples

We can create an experiment by just passing the name, and the client is inferred (respects `mlflow_set_tracking_uri()` and the `MLFLOW_TRACKING_URI` environment variable.)
```
experiment <- mlflow_create_experiment("myexp")
experiment

# A tibble: 1 x 4
  experiment_id name  artifact_location                     lifecycle_stage
  <chr>         <chr> <chr>                                 <chr>          
1 1             myexp /Users/kevinykuo/Dropbox/Projects/ml… active    
```

Similarly, we can create a run by providing the experiment ID.

```
experiment_id <- experiment$experiment_id
run <- mlflow_create_run(experiment$experiment_id)
dplyr::glimpse(run)

Observations: 1
Variables: 13
$ run_uuid        <chr> "a15b485c7fd0482c8b6cce302c581196"
$ experiment_id   <chr> "1"
$ name            <chr> ""
$ source_type     <chr> "NOTEBOOK"
$ source_name     <chr> ""
$ user_id         <chr> "kevinykuo"
$ status          <chr> "RUNNING"
$ start_time      <dttm> 2019-04-12 02:37:35
$ artifact_uri    <chr> "/Users/kevinykuo/Dropbox/Projects/mlflow-ex/mlruns/1/a15b485c7fd0482c8…
$ lifecycle_stage <chr> "active"
$ metrics         <lgl> NA
$ params          <lgl> NA
$ tags            <lgl> NA
```

Log some metrics and inspect.

```
run_id <- run$run_uuid
mlflow_log_metric("mse", 24, run_id = run_id)
mlflow_log_metric("mse", 25, run_id = run_id)

mlflow_get_metric_history(run_id, "mse")

# A tibble: 2 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 mse      24 2019-04-12 02:38:47
2 mse      25 2019-04-12 02:38:47
```

We can also explicitly pass a client.

```
client <- mlflow_client()

mlflow_log_metric("mse", 26, client = client, run_id = run_id)
mlflow_get_metric_history(client = client, run_id, "mse")

# A tibble: 3 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 mse      24 2019-04-12 02:38:47
2 mse      25 2019-04-12 02:38:47
3 mse      26 2019-04-12 02:39:21
```

When only `client` is provided, we require a `run_id` to avoid ambiguity.

```
mlflow_log_metric("mse", 27, client = client)
## Error: `run_id` must be specified when `client` is specified. 
```

Each `RunInfo` is represented as a row in the tibble.

```
mlflow_list_run_infos(experiment$experiment_id)
# A tibble: 1 x 10
  run_uuid experiment_id name  source_type source_name user_id status start_time         
  <chr>    <chr>         <chr> <chr>       <chr>       <chr>   <chr>  <dttm>             
1 a15b485… 1             ""    NOTEBOOK    ""          keviny… RUNNI… 2019-04-12 02:37:35
# … with 2 more variables: artifact_uri <chr>, lifecycle_stage <chr>
```

For batch logging, we take named lists, which is more natural than requiring nested lists with `key` and `value`. An optional param `timestamps` can be used if metrics need to be associated with times that are not the current time.
```
run <- mlflow_create_run(experiment$experiment_id)
mlflow_log_batch(
  metrics = list(rmse = 24, auc = 0.99), 
  params = list(alpha = 0.5, lambda = 0.5),
  tags = list(foo = "bar", bar = "baz"),
  run_id = run$run_uuid
)
```

Again, each run is a row:

```
run <- mlflow_get_run(run$run_uuid)

> dplyr::glimpse(run)
Observations: 1
Variables: 13
$ run_uuid        <chr> "0abd954521a94bf095569250a669dfcb"
$ experiment_id   <chr> "1"
$ name            <chr> ""
$ source_type     <chr> "NOTEBOOK"
$ source_name     <chr> ""
$ user_id         <chr> "kevinykuo"
$ status          <chr> "RUNNING"
$ start_time      <dttm> 2019-04-12 02:40:18
$ artifact_uri    <chr> "/Users/kevinykuo/Dropbox/Projects/mlflow-ex/mlruns/1/0abd954521a94bf09…
$ lifecycle_stage <chr> "active"
$ metrics         <list> [<tbl_df[2 x 3]>]
$ params          <list> [<tbl_df[2 x 2]>]
$ tags            <list> [<tbl_df[2 x 2]>]


run$metrics

[[1]]
# A tibble: 2 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 auc    0.99 2019-04-12 02:40:18
2 rmse  24    2019-04-12 02:40:18
```

kevinykuo committed:
export mlflow_client() and rebuild docs

kevinykuo committed:
reimport httr functions

kevinykuo committed:
add mlflow_list_experiments()

kevinykuo committed:
add logs/ to .Rbuildignore

kevinykuo committed:
refactor mlflow_log_metric()

kevinykuo committed:
add `call. = FALSE` to mlflow_rest errors

kevinykuo committed:
refactor mlflow_create_experiment()

kevinykuo committed:
refactor mlflow_log_metric() to not use S3

kevinykuo committed:
refactor mlflow_list_experiments()

kevinykuo committed:
refactor mlflow_create_experiment()

kevinykuo committed:
add mlflow_get_experiment()

kevinykuo committed:
refactor mlflow_list_experiments() and add print method

kevinykuo committed:
import zeallot and tibble

kevinykuo committed:
add mlflow_delete_experiment()

kevinykuo committed:
add mlflow_restore_experiment()

kevinykuo committed:
add mlflow_rename_experiment()

kevinykuo committed:
create delete and restore run

kevinykuo committed:
get run

kevinykuo committed:
call mlflow_create_run in mlflow_start_run

kevinykuo committed:
add mlflow_log_batch

kevinykuo committed:
refactor mlflow_set_tag()

kevinykuo committed:
refactor mlflow_log_param()

kevinykuo committed:
add get metric history

kevinykuo committed:
add search runs

kevinykuo committed:
refactor list artifacts

kevinykuo committed:
add mlflow_set_terminated

kevinykuo committed:
add mlflow_download_artifacts()

kevinykuo committed:
refactor reset array print method

kevinykuo committed:
add mlflow_get_experiment_by_name

kevinykuo committed:
restore/delete/rename methods return tibble

kevinykuo committed:
mlflow_get_run returns tibble

kevinykuo committed:
restore run returns tibble

kevinykuo committed:
refactor mlflow_set_terminated

kevinykuo committed:
remove tracking rest data structures

kevinykuo committed:
update NAMESPACE

kevinykuo committed:
use parse_run in mlflow_create_run()

kevinykuo committed:
mlflow_search_runs() returns tibble

kevinykuo committed:
add mlflow_list_run_infos()

kevinykuo committed:
allow null search criterion for mlflow_search_runs()

kevinykuo committed:
refactor artifacts logging/listing

kevinykuo committed:
logging artifact returns artifact listing

kevinykuo committed:
mlflow_create_experiment() returns tibble

kevinykuo committed:
set tag returns NULL

kevinykuo committed:
fix tracking tests

kevinykuo committed:
fix lints

kevinykuo committed:
add file:// to schemeless uri

kevinykuo committed:
fix mlflow_start_run()

kevinykuo committed:
add a comment

kevinykuo committed:
r cmd check fixes

kevinykuo committed:
workaround databricks test

kevinykuo committed:
rename files

kevinykuo committed:
remove rest wrappers

kevinykuo committed:
fix mlflow_download_artifacts() signature

kevinykuo committed:
rebuild docs

kevinykuo committed:
cast experiment_id to string

tomasatdatabricks reviewed (commented):
Looks good, thanks Kevin!

a) I have a couple of high level questions:
1.  mlflow_start_run vs mlflow_create_run - can / should we just keep one of them? mlflow_start_run?
2. Do we want to use active experiment / run id to fill in the arguments whenever possible?
    - I think this would be ok and simplify the behavior except for delete / restore calls. I wonder if we can just treat these as special case and throw exception if you are attempting to delete active run / experiment?
3. [minor] For all the apis that use mlflow_client = NULL, should we delay the creation of the client until it's actually needed? I.e. just pass the client or NULL along.

b) It loos like there is a broken merge in tracking-rest.R, I left a comment in code.

kevinykuo committed:
parse_run() returns tbl_mlflow_run and implement with.tbl_mlflow_run

kevinykuo committed:
endow experiments with tbl_mlflow_experiment class

kevinykuo committed:
rename mlflow_get_active_experiment_id() to mlflow_active_experiment_id()

kevinykuo committed:
error if user tries to delete active experiment

kevinykuo committed:
make experiment_id optional in mlflow_rename_experiment()

kevinykuo committed:
resolve_experiment_id should actually return the id

kevinykuo committed:
experiment_id optional in search runs and list run infos

kevinykuo committed:
rename mlflow_get_active_run()

kevinykuo committed:
rename mlflow_get_active_experiment_id()

kevinykuo committed:
rename active_run_id() to mlflow_get_active_run_id()

kevinykuo committed:
rename mlflow_get_run_id()

kevinykuo committed:
prevent deletion of active run

kevinykuo committed:
run_id optional in mlflow_get_run()

kevinykuo committed:
run_id optional in mlflow_get_metric_history()

kevinykuo committed:
options run_id in mlflow_list_artifacts and mlflow_set_terminated

kevinykuo committed:
optional run_id in mlflow_download_artifacts()

kevinykuo committed:
consistent param ordering

kevinykuo committed:
fix failed git-fu

kevinykuo committed:
typo

kevinykuo committed:
fixes

kevinykuo committed:
allow looking at env variables for experiment id

kevinykuo committed:
name output of resolve_client_and_run_id()

kevinykuo committed:
don't use 'tbl' in mlflow run and experiment class names

kevinykuo committed:
class names in client constructors

kevinykuo committed:
dispatch mlflow_experiment

kevinykuo committed:
experiment_id optional in mlflow_create_run()

kevinykuo committed:
fix tags checking

kevinykuo committed:
remove entities.R

kevinykuo committed:
rename mlflow_get_active_experiment()

kevinykuo committed:
add mlflow_id()

kevinykuo committed:
use identical() to check for active expeirment id in mlflow_delete_experiment()

kevinykuo committed:
tests for experiments tracking

kevinykuo committed:
absorb mlflow_get_experiment/by_name

kevinykuo committed:
mlflow_set_experiment takes both name and id and returns the experiment

kevinykuo committed:
fix tests

kevinykuo committed:
mlflow_get_active_run_id() shouldn't need to call mlflow_get_active_run()

kevinykuo committed:
don't use mlflow_get_active_run() in mlflow_set_experiment() and mlflow_end_run()

kevinykuo committed:
get active run id instead of active run

kevinykuo committed:
cache run id instead of run object

kevinykuo committed:
remove active run getter

kevinykuo committed:
don't cache active experiment

kevinykuo committed:
mlflow_start_run() should return a run

kevinykuo committed:
fix test

kevinykuo committed:
reimplement mlflow_active_run()

kevinykuo committed:
add test for logging

kevinykuo commented on mlflow/R/mlflow/R/tracking-methods.R:
```diff
@@ -0,0 +1,602 @@
+#' @template roxlate-run-id
+#' @template roxlate-client
+#' @export
+mlflow_log_metric <- function(key, value, timestamp = NULL, client = NULL, run_id = NULL) {
```
Done!

kevinykuo commented on mlflow/R/mlflow/R/tracking-rest.R:
```diff
@@ -53,6 +54,10 @@ mlflow_rest <- function( ..., client, query = NULL, data = NULL, verb = "GET", v
     paste(args, collapse = "/")
   )
+
+  config <- config(ssl_verifypeer = rest_config$verify_peer)
```
Failed git-fu on my part, should be fixed.

kevinykuo commented on mlflow/R/mlflow/R/tracking-utils.R:
```diff
@@ -101,3 +101,41 @@ MLFLOW_SOURCE_TYPE <- list(
+  } else {
+    if (is.null(run_id)) stop("`run_id` must be specified when `client` is specified.", call. = FALSE)
+  }
+  list(client, run_id)
```
OK

kevinykuo commented on mlflow/R/mlflow/R/tracking-client.R:
```diff
@@ -114,381 +117,10 @@ new_mlflow_client.mlflow_https <- function(tracking_uri) {
 mlflow_client <- function(tracking_uri = NULL) {
   tracking_uri <- new_mlflow_uri(tracking_uri %||% mlflow_get_tracking_uri())
   client <- new_mlflow_client(tracking_uri)
   mlflow_validate_server(client)
```
Updated to validate only for file client

kevinykuo commented:
@tomasatdatabricks Thanks! Comments have been addressed and some thoughts re: your questions below.
- On `mlflow_create_run()` and `mlflow_start_run()` -- I think we should keep both. Users are used to the latter starting an *active* run. Also the Python docs has been [explicit in saying that "create run" does not register the active run](https://mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.create_run).
- We'll fill in active experiment/run IDs whenever possible. Informative errors are given when user tries to delete an active experiment.
- The client is always needed because we always hit the REST API.

tomasatdatabricks commented:
Hey Kevin, thanks for the updates addressing my comments!

I have some more questions regarding mlflow_create_run / mlflow_start_run. 

a) Personally I think it's a  bit weird that mfllow_start_run infers some information from the context by calling mlflow_get_run_context, whereas create run does not. I.e. the difference between the two is not just whether the run is started / activated. I don't think we should be calling get_run_context in create run since create run is supposed to give user full control over what is being recorded. The same is true in python. However, python client has a clear distinction between client and fluent apis and therefore it's a bit more obvious that these are different functions. Is there a good way how we can make this distinction more obvious in R? Maybe we can refactor the code so that get_run_context is called inside mlflow_create_run but user can override it by passing in context.

b) Can we somehow make it clear that you do not have call mlflow_create_run in order to start one? I am worried people might understand it that way.

I would consider grouping the more lower level apis together somehow. E.g. with mlflow_manage prefix. I think it would be easier to explain to people that mlflow_manage is not meant to be used on regular basis.

What do you think?

kevinykuo committed:
unexport mlflow_create_run()

kevinykuo committed:
absorb mlflow_create_run params into mlflow_start_run()

kevinykuo committed:
update test

kevinykuo committed:
fix mlflow_source() cleanup code

kevinykuo committed:
absorb mlflow_set_terminated into mlflow_end_run()

kevinykuo committed:
reorganize and add mlflow_active_experiment()

kevinykuo committed:
fix docs

kevinykuo commented:
Removed both `mlflow_create_run()` and `mlflow_set_terminated()` in favor of `mlflow_start/end_run()`.

tomasatdatabricks commented on mlflow/R/mlflow/R/tracking-runs.R:
```diff
@@ -0,0 +1,597 @@
+
+  active_run_id <- mlflow_get_active_run_id()
+
+  if (!is.null(client) && is.null(run_id))
```
should we just call resolve_client_and_run_id?

kevinykuo replied:
`resolve_client_and_run_id()` could start a run though

tomasatdatabricks commented on mlflow/R/mlflow/R/tracking-runs.R:
```diff
@@ -0,0 +1,597 @@
+#'
+#' @examples
+#' \dontrun{
+#' with(mlflow_start_run(), {
```
should this be with.mlflow_run?

kevinykuo replied:
`mlflow_start_run()` returns an `mlflow_run` object so `with` will dispatch accordingly

tomasatdatabricks commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+#' @template roxlate-client
+#' @export
+mlflow_get_experiment <- function(name = NULL, experiment_id = NULL, client = NULL) {
+  if (!is.null(name) && !is.null(experiment_id)) {
```
should we attempt to infer the experiment if no id nor name is provided?

kevinykuo replied:
yeah I think it makes sense to return the active experiment or whatever the user set in the env variable?

tomasatdatabricks replied:
Yeah, sounds good to me.

tomasatdatabricks commented on mlflow/R/mlflow/R/tracking-client.R:
```diff
@@ -4,6 +4,9 @@ new_mlflow_client <- function(tracking_uri) {
 
 new_mlflow_uri <- function(raw_uri) {
+  if (!grepl("://", raw_uri)) {
+    raw_uri <- paste0("file://", raw_uri)
```
we need special case for "databricks".

tomasatdatabricks commented on mlflow/R/mlflow/tests/testthat/test-databricks-utils.R:
```diff
@@ -14,7 +14,7 @@ test_that("mlflow creates databricks client when scheme is databricks", {
       list(experiments = c(1, 2, 3))
     }, {
-      client1 <- mlflow:::mlflow_client("databricks")
+      client1 <- mlflow:::mlflow_client("databricks://")
```
This needs to remain "databricks".

kevinykuo replied:
Will add special case for this.

tomasatdatabricks reviewed (changes requested):
The PR looks really good, thanks Kevin!

I had a few small comments / questions in code  + the following:

1. We need a special case for tracking_uri == "databricks". 
2. I did not see any tests for mlflow_log_batch? We need to add some or take the log_batch out and make a follow up PR.

I still want to double check test coverage for the rest.

kevinykuo commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+}
+
+mlflow_get_experiment_by_name <- function(name, client = NULL) {
+  client <- client %||% mlflow_client()
```
Ideally we should check all user inputs.

kevinykuo commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  if (is.null(experiment_name) && is.null(experiment_id)) {
+    stop("At least one of `experiment_name` or `experiment_id` should be specified.",
```
Will fix.

kevinykuo commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+    verb = "POST", client = client,
+    data = list(experiment_id = experiment_id)
+  )
+  mlflow_get_experiment(experiment_id = experiment_id)
```
We're generally following the rule that operations on experiments return experiments, operations on runs return runs, (and logging operations are currently inconsistent.) I think it makes sense to give the user feedback right away on the side effect, and if we're printing it we might as well return it.

tomasatdatabricks replied:
Ok, sounds good for now but I think we should revisit this before 1.0. There can be some subtle differences in behavior compared to python.

kevinykuo commented:
Will add a test for `mlflow_log_batch()`.

kevinykuo commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+#' Set Experiment
+#'
+#' Sets an experiment as the active experiment. If the experiment does not exist,
+#'   creates an experiment with provided name.
```
Will fix.

kevinykuo commented on mlflow/R/mlflow/R/tracking-experiments.R:
```diff
@@ -0,0 +1,209 @@
+#' Retrieves the active experiment.
+#'
+#' @export
+mlflow_active_experiment <- function() {
```
If we have `mlflow_active_run()` it's reasonable to assume we have `mlflow_active_experiment()`? One question here is whether a provided `MLFLOW_EXPERIMENT_NAME/ID` should be returned if an experiment hasn't been "set" (via `mlflow_set_experiment()`).

kevinykuo commented on mlflow/R/mlflow/R/tracking-globals.R:
```diff
@@ -1,27 +1,19 @@
-mlflow_active_run <- function() {
-  .globals$active_run
-}
+mlflow_get_run_id <- function(run) cast_nullable_string(run$run_uuid)
```
This function isn't used, will remove.

tomasatdatabricks replied:
@kevinykuo can you remove it?

kevinykuo commented on mlflow/R/mlflow/R/tracking-runs.R:
```diff
@@ -0,0 +1,597 @@
+
+
+
+
```
OK

kevinykuo committed:
remove mlflow_active_run()

kevinykuo committed:
remove mlflow_active_experiment()

kevinykuo committed:
add test for log batch

kevinykuo committed:
special-case 'databricks' uri

kevinykuo committed:
better error message when both experiment_name and experiment_id are provided

kevinykuo committed:
clean up input checking

kevinykuo committed:
improve set experiment and start run docs

kevinykuo committed:
attempt to fix flaky log_batch test

kevinykuo commented:
@tomasatdatabricks 
- Removed `mlflow_active_run()` and `mlflow_active_experiment()` in favor of `mlflow_get_run()` and `mlflow_get_experiment()`. 
- Special case `databricks` URI so `://` isn't needed.
- Added tests for `mlflow_log_batch()`.
- Added suggested docs/error message improvements.

kevinykuo commented:
@tomasatdatabricks can we merge?

tomasatdatabricks reviewed (changes requested):
There is one last nit that needs to be resolved, after that we can merge.

kevinykuo committed:
remove unused mlflow_get_run_id()

kevinykuo commented:
Oops missed that one, fixed!

tomasatdatabricks reviewed (approved):
lgtm!
